### DISCUSSION TITLE
CREXX performance comparison

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#40] [General]
I did a performance test with a little rexx 

say time('l')
do i=1 to 1000000000
   b=i-1
end
say time('l')

calculated times:

plain C:     1.071401 seconds 
CREXX:      8             seconds  
BREXX:    74.106230 seconds  
Regina  156.125000 seconds   (loop just up to 950 million, above it crashed)

maybe an unfair comparison ...

**Comment by rvjansen** [comment]
Interesting. It prompts the question what kind of supercomputer you have  - or if we had some performance regression, because on a not too shabby macbook pro from 2017, this clocks at 27 sec, and the raspberry and the jetson at 100 sec, and 2 minutes. This is what I run:

/* REXX COMPILER PoC */

.globals=0

main()   .locals=7
   time r6
   say r6
   say "\n"
   * Line 2: {DO}
   * Line 2: i=1
   load r2,1
   * Line 2: to 1000000000
   load r3,1000000000
l20:
   * Line 2: to 1000000000
   igt r0,r2,r3
   brt l13,r0
   * Line 3: b=i-1
   load r5,1
   isub r4,r2,r5
   * Line 2: {Implicit "BY 1"}
   inc r2
   * Line 2: {DO-END}
   br l20
l13:
   time r6
   say r6
   say "\n"
   exit

**Reply by adesutherland** [reply]
Debug / Release? 

In the attached test (which is basically the same as yours) I get:

    rxvm perf1 - debug
    Execution Time: 51
    rxvm perf1 - release
    Execution Time: 10

[perf1.zip](https://github.com/adesutherland/CREXX/files/6680242/perf1.zip)

I will start adding performance tests to the tests folder because you are right - we need to keep on top of this!

**Reply by adesutherland** [reply]
```
Windows 10
Processor	Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz   2.71 GHz
Installed RAM	8.00 GB (7.88 GB usable)
System type	64-bit operating system, x64-based processor
```

**Comment by adesutherland** [comment]
Can you try in branch f0017?

A

**Reply by mgrossmann** [reply]
Sorry @adesutherland,  it's simply to hot over here.
After assembling all went fine. 👯 


@rvjansen did you build the make file in this way?
`➜  CREXX-build cmake -DCMAKE_BUILD_TYPE=Release ../CREXX/`


On develop:
```
➜  interpreter ./rxvm perf1 
Execution Time: 6
```

On feature/f0017:
```
➜  interpreter ./rxvm perf1       
Execution Time: 7
```


My system:
```
  Modellname:	iMac (late 2013)
  Modell-Identifizierung:	iMac14,2
  Prozessortyp:	Quad-Core Intel Core i7
  Prozessorgeschwindigkeit:	3,5 GHz
  Anzahl der Prozessoren:	1
  Gesamtanzahl der Kerne:	4
  L2-Cache (pro Kern):	256 KB
  L3-Cache:	8 MB
  Hyper-Threading Technologie:	Aktiviert
  Speicher:	16 GB
```

**Comment by adesutherland** [comment]
I changed the shell script to also test bytecode mode:

```
#!/bin/sh
echo "Performance Test 1 "

echo "rxas perf1"
../cmake-build-release/assembler/rxas perf1

echo "rxdas perf1"
../cmake-build-release/disassembler/rxdas perf1

echo "rxvm perf1 - debug"
../cmake-build-debug/interpreter/rxvm perf1

echo "rxvm perf1 - release"
../cmake-build-release/interpreter/rxvm perf1

echo "rxvm perf1 - bytecode mode - debug"
../cmake-build-debug/interpreter/rxbvm perf1

echo "rxvm perf1 - bytecode mode - release"
../cmake-build-release/interpreter/rxbvm perf1
```

And get 
```
rxvm perf1 - debug
Execution Time: 41
rxvm perf1 - release
Execution Time: 8
rxvm perf1 - bytecode mode - debug
Execution Time: 41
rxvm perf1 - bytecode mode - release
Execution Time: 16
```

Which is interesting actually

A

**Comment by rvjansen** [comment]
Mike,

You have a point there - I assumed the standard build was Release and I needed to specify Debug when I wanted that - so no optimization is going on. Will try that after morning coffee.

René.

**Comment by rvjansen** [comment]
*BOOM!*
➜  general git:(feature/f0004) ✗ time rxvm time_loop
1624107370
1624107376
rxvm time_loop  6.38s user 0.00s system 99% cpu 6.387 total

The time instruction can use more granularity.

That optimization in C is amazing! For me now the final reason to switch to out-of-source builds, so I can have a debug and optimized version on each box.
Now looking for a good set of aliases to switch quickly.

(And it was already so fast!)

Adrian, if you can put a ‘performance’ subdir in tests, and put your benchmarks in there, it will merge nicely.

btw, I am looking for the windows equivalent for:

find . -name "*.rexx" | awk '{$$2 = $$1 ; sub ( /\.rexx/,".rxas",$$1 ) ; print $$1 }' | xargs rm -f

for use in Cmake, so I can keep .rxas around for which there is no .rexx, and still clean .rxas for when there is.

best regards,

René.

**Comment by adesutherland** [comment]
>  That optimization in C is amazing!

I think what we are seeing is a lot of inlining of functions (the vm code
is structured to point the compiler in that direction).

Are you using clang or gcc?

And also this is why I still want to see if we can convert REXX Assembly to
LLVM IL and use all the smart optimisations that those guys have done.

A

**Comment by rvjansen** [comment]
It tells me it is clang-1205.0.22.9

René.

**Comment by rvjansen** [comment]
This shows the difference between rxvm and rxbvm:
➜  general git:(feature/f0004) ✗ perf stat -d rxvm time_loop 
1625261065
1625261089

 Performance counter stats for 'rxvm time_loop':

          23391.18 msec task-clock                #    0.999 CPUs utilized          
               398      context-switches          #    0.017 K/sec                  
                 2      cpu-migrations            #    0.000 K/sec                  
                45      page-faults               #    0.002 K/sec                  
       35079814541      cycles                    #    1.500 GHz                    
       66067475740      instructions              #    1.88  insn per cycle         
   <not supported>      branches                                                    
            191338      branch-misses                                               
       35031206432      L1-dcache-loads           # 1497.624 M/sec                  
            272820      L1-dcache-load-misses     #    0.00% of all L1-dcache hits  
   <not supported>      LLC-loads                                                   
   <not supported>      LLC-load-misses                                             

      23.404026746 seconds time elapsed

      23.389218000 seconds user
       0.004000000 seconds sys


➜  general git:(feature/f0004) ✗ perf stat -d rxbvm time_loop 
1625261113
1625261162

 Performance counter stats for 'rxbvm time_loop':

          49447.06 msec task-clock                #    0.999 CPUs utilized          
              1373      context-switches          #    0.028 K/sec                  
                 4      cpu-migrations            #    0.000 K/sec                  
                46      page-faults               #    0.001 K/sec                  
       74157882404      cycles                    #    1.500 GHz                    
      110146737336      instructions              #    1.49  insn per cycle         
   <not supported>      branches                                                    
            458853      branch-misses                                               
       50068138207      L1-dcache-loads           # 1012.560 M/sec                  
            813999      L1-dcache-load-misses     #    0.00% of all L1-dcache hits  
   <not supported>      LLC-loads                                                   
   <not supported>      LLC-load-misses                                             

      49.488221297 seconds time elapsed

      49.437678000 seconds user
       0.011992000 seconds sys
